subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultWatchableFileSystemDetector.java

```diff
@@ -0,0 +1,92 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal.watch.vfs.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import net.rubygrapefruit.platform.Native;
+import net.rubygrapefruit.platform.file.FileSystemInfo;
+import net.rubygrapefruit.platform.file.FileSystems;
+import org.gradle.internal.watch.vfs.WatchableFileSystemDetector;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.nio.file.Path;
+import java.util.stream.Stream;
+
+public class DefaultWatchableFileSystemDetector implements WatchableFileSystemDetector {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWatchableFileSystemDetector.class);
+
+    // !IMPORTANT! If changed, make sure to update the documentation in gradle_damon.adoc
+    private static final ImmutableSet<String> SUPPORTED_FILE_SYSTEM_TYPES = ImmutableSet.of(
+        // APFS on macOS
+        "apfs",
+        // HFS and HFS+ on macOS
+        "hfs",
+        "ext3",
+        "ext4",
+        "btrfs",
+        // NTFS on macOS
+        "ntfs",
+        // NTFS on Windows
+        "NTFS",
+        // FAT32 on macOS
+        "msdos",
+        // FAT32 on Windows
+        "FAT32",
+        // exFAT on macOS
+        "exfat",
+        // exFAT on Windows
+        "exFAT",
+        // VirtualBox FS
+        "vboxsf"
+    );
+
+    private final FileSystems fileSystems;
+
+    public static WatchableFileSystemDetector create() {
+        return new DefaultWatchableFileSystemDetector(Native.get(FileSystems.class));
+    }
+
+    @VisibleForTesting
+    DefaultWatchableFileSystemDetector(FileSystems fileSystems) {
+        this.fileSystems = fileSystems;
+    }
+
+    @Override
+    public Stream<FileSystemInfo> detectUnsupportedFileSystemsUnder(Path root) {
+        return fileSystems.getFileSystems().stream()
+            .filter(fileSystem -> {
+                LOGGER.debug("Detected {}: {} from {} (remote: {}, case-sensitive: {}, case-preserving: {})",
+                    fileSystem.getFileSystemType(),
+                    fileSystem.getMountPoint(),
+                    fileSystem.getDeviceName(),
+                    fileSystem.isRemote(),
+                    fileSystem.isCaseSensitive(),
+                    fileSystem.isCasePreserving());
+
+                // We don't support network file systems
+                if (fileSystem.isRemote()) {
+                    return true;
+                }
+                if (SUPPORTED_FILE_SYSTEM_TYPES.contains(fileSystem.getFileSystemType())) {
+                    return false;
+                }
+                return root.startsWith(fileSystem.getMountPoint().toPath());
```

Comment:
This method doesn't seem to do what I expect it to do. It will return:
- All remote file systems
- The file systems with an unsupported type which are a parent of the given path.

I think what it should return instead is:
- Then unsupported file systems which are children of the root path
- The closest (aka longest) file system which is a parent of the root path, if watching for it is unsupported.

Reply:
I fixed this now. I think it should return any unsupported parent. I don't think we should try to enable file system watching on a supported file system if it's mounted under an unsupported one, so the rules should be:

* The unsupported file systems which are children of the root path
* The unsupported file systems which are parents of the root path

Reply:
Actually, we could simply this further now that the logic is about invalidation. We could simply invalidate all unsupported file systems every time a build finishes. We have no business retaining data from any of those file systems, and doing so will be a no-op for the happy day case.

Reply:
FTR, did this.